tdp/core/dag.py

```diff
@@ -152,51 +154,100 @@ def graph(self, value: nx.DiGraph) -> None:
     def graph(self) -> None:
         self.graph = None
 
-    def topological_sort(
-        self, nodes: list[str] = None, restart: bool = False
-    ) -> list[Operation]:
-        """Perform a topological sort on the DAG.
+    def node_to_operation(self, node: str, restart: bool = False) -> Operation:
+        if restart and node.endswith("_start"):
+            node = node.replace("_start", "_restart")
+        return self.collections.get_operation(node)
 
-        Args:
-            nodes: List of nodes to sort.
-            restart: If True, restart operations are returned instead of start operations.
+    def topological_sort_key(
+        self,
+        items: Optional[Iterable[T]] = None,
+        key: Optional[Callable[[T], str]] = None,
+    ) -> Generator[T, None, None]:
+        """Perform a topological sort on the DAG with an iterable.
 
-        Returns:
-            List of operations sorted topologically.
+        A "key" function can be supplied if the "items" iterable is not a DAG node.
+        The "key" function must return a valid DAG node with element of "items" iterable as input.
+
+        Example:
+            dag = Dag(...)
+            items = [("hdfs_start", "foo"), ("hdfs_config", "bar")]
+            items = dag.topological_sort_key(items, key=lambda x: x[0])
+            # items = [("hdfs_config", "bar"), ("hdfs_start", "foo")]
+
+        Args:
+            items: Iterable to sort.
+            key: Function to map "items" element to DAG node.
         """
```

Comment:
Done but I updated the first sentence because the lambda is removed.
```diff
- By default, if no "key" function is provided, each item is converted to string format to match against DAG nodes.
+ By default, if no "key" function is provided, each item is used as is and must match against DAG nodes.
```